notebooks/generative/fashionmnist_vae.livemd

```diff
@@ -0,0 +1,526 @@
+# A Variational Autoencoder for MNIST
+
+```elixir
+Mix.install([
+  {:exla, "~> 0.3.0"},
+  {:nx, "~> 0.3.0"},
+  {:axon, "~> 0.2.0"},
+  {:req, "~> 0.3.1"},
+  {:kino, "~> 0.6.2"},
+  {:scidata, "~> 0.1.9"},
+  {:stb_image, "~> 0.5.2"},
+  {:kino_vega_lite, "~> 0.1.3"},
+  {:vega_lite, "~> 0.1.6"}
+])
+
+alias VegaLite, as: Vl
+```
+
+## Introduction
+
+In this notebook, we'll be building a variational autoencoder (VAE). This will help demonstrate splitting up models, defining custom layers and loss functions, using multiple outputs, and a few additional Kino tricks for training models.
+
+This notebook builds on the [denoising autoencoder example](mnist_autoencoder_using_kino.livemd) and turns the simple autoencoder into a variational one for the same dataset.
+
+## Training a simple autoencoder
+
+This section will proceed without much explanation as most of it is extracted from [denoising autoencoder example](mnist_autoencoder_using_kino.livemd). If anything here doesn't make sense, take a look at that notebook for an explanation.
+
+```elixir
+defmodule Data do
+  @moduledoc """
+  A module to hold useful data processing utilities, 
+  mostly extracted from the previous notebook
+  """
+
+  @doc """
+  Converts the given image into a `Kino.Image`.
+
+  `image` must be a single channel `Nx` tensor with pixel values between 0 and 1.
+  `height` and `width` are the output size in pixels
+  """
+  def image_to_kino(image, height \\ 200, width \\ 200) do
+    image
+    |> Nx.multiply(255)
+    |> Nx.as_type(:u8)
+    |> Nx.transpose(axes: [:height, :width, :channels])
+    |> StbImage.from_nx()
+    |> StbImage.resize(height, width)
+    |> StbImage.to_binary(:png)
+    |> Kino.Image.new(:png)
+  end
+
+  @doc """
+  Converts image data from `Scidata.MNIST` into an `Nx` tensor and normalizes it.
+  """
+  def preprocess_data(data) do
+    {image_data, _labels} = data
+    {images_binary, type, shape} = image_data
+
+    images_binary
+    |> Nx.from_binary(type)
+    # Since pixels are organized row-wise, reshape into rows x columns
+    |> Nx.reshape(shape, names: [:images, :channels, :height, :width])
+    # Normalize the pixel values to be between 0 and 1
+    |> Nx.divide(255)
+  end
+
+  @doc """
+  Converts a tensor of images into paired, batched images for model training
+  """
+  def prepare_training_data(images, batch_size) do
+    batched_images = images |> Nx.to_batched(batch_size)
+    Stream.zip(batched_images, batched_images)
```

Comment:
@seanmor5 @jonatanklosko Is there a way to make these batches random? Doing `images |> Nx.shuffle |> Nx.to_batched(batch_size)` doesn't do the right thing since that only shuffles it once. I want the shuffling to be different for every epoch? I couldn't figure out how to do this - I think it probably has something to do with `Stream.resource/3` but couldn't figure out how to make that work without a `Stream.next` function.

I think it'd be useful if `Nx.to_batched` took a `randomize` option

Reply:
I came up with this but it seems a little messy:

```elixir
shuffled_batches = Stream.resource(
  fn -> 
    batches = Nx.iota({100, 1}) |> Nx.shuffle |> Nx.to_batched(5)
    {0, batches}
  end,
  fn {i, batches} -> 
    case Enum.at(batches, i) do
      nil -> {:halt, {i, batches}}
      value -> {[value], {i+1, batches}}
    end
  end,
  fn _acc -> nil end
)
data = Stream.map(shuffled_batches, fn batch -> {batch, batch} end)
```

Reply:
You could use `Enum.shuffle` but I think that also would only shuffle it once. I can probably add something to `Axon.Loop` that shuffles the enumerated dataset every epoch

Reply:
@hanrelan we need a stream where all images are shuffled on every run, so I'm thinking this:

```elixir
Stream.flat_map([images], fn images ->
  images |> Nx.shuffle() |> Nx.to_batched(batch_size)
end)
```

In fact, it could be `Stream.flat_map([:whatever], fn _ -> ... end)` :)

Reply:
Hmm I tried this @jonatanklosko and it became really slow to shuffle. If I tried putting it into a `defn` it complains that `to_batched` isn't allowed within `defn`. Will just leave it unshuffled for now and leave this for another day :)

Reply:
Ah, you are only calling EXLA compiler explicitly in the notebook, so regular operations are done using the binary backend. Try adding `Nx.global_default_backend(EXLA.Backend)` below mix install and see if that speeds it up.

Reply:
That worked! I also needed to specify `axis: :images` on the `Nx.shuffle` so it didn't shuffle everything. Specifying the axis sped it up to about a second, specifying the global backend made it feel instant after that.